### ISSUE TITLE
OSDI 26 ｜ Tessera: A Holistic Pipeline Parallelism Framework for Trillion-Parameter Heterogeneous MoE Training

### ISSUE THREAD

**Issue by woshipapa** [AI-Infra-Team/awesome-papers#218]
Title: Tessera: A Holistic Pipeline Parallelism Framework for Trillion-Parameter Heterogeneous MoE Training (Operational Systems)
Authors: Weifang Hu, Langshi Chen, Man Yuan, Youyang Yao, Xiulong Yuan, Li Tian, Yong Li, Wei Lin, Xuanhua Shi, Zhengping Qian, Jingren Zhou
Institutes: Huazhong University of Science and Technology; Alibaba Cloud
Paper: osdi26-hu-weifang.pdf
Conference Page: https://www.usenix.org/conference/osdi26/presentation/hu-weifang

<img width="470" height="290" alt="Image" src="https://github.com/user-attachments/assets/a7f9d966-3c3f-4637-8bcd-b9e8f90fb610" />

<img width="453" height="536" alt="Image" src="https://github.com/user-attachments/assets/30b2aa38-36ba-4036-b36e-ce262f5a1873" />

---
1 Background and Observation

Trillion-parameter training relies on pipeline parallelism, which splits layers into chunks assigned to pipeline stages. The interleaved 1F1B schedule gives each rank several virtual stages to reduce bubbles. For MoE models, the dominant communication cost is the All-to-All (A2A) used for token dispatch and combine.

Tessera hides A2A through inter-microbatch overlap: in an interleaved schedule, adjacent operations on a rank belong to different chunks and different microbatches, so they carry no data dependency and can run concurrently — one chunk's A2A hides under another chunk's computation. Two such co-scheduled operations form an overlap pair, measured by its post-overlap cost (the makespan of the pair), not the serial sum of its parts.

Existing pipeline systems rest on a uniformity assumption: the model is a stack of identical Transformer bricks, so every overlap pair is structurally the same and one hand-crafted overlap template suffices.

Qwen3-Next breaks this. It interleaves Gated DeltaNet (GDN) linear-attention layers with standard softmax-attention layers in a 3:1 pattern, each followed by a sparse MoE FFN. At long sequence lengths this creates up to 10× compute asymmetry between adjacent attention layers, so different chunk combinations present structurally different compute–communication timelines. Measured overlap gain varies by 3×.

---
2 Core Problem and Opportunity

<img width="428" height="385" alt="Image" src="https://github.com/user-attachments/assets/4206668a-1496-4545-bdff-77b517972fd9" />

(a) Combinatorial overlap space. With K ranks and vp virtual stages each, a rank's co-resident chunks form K · vp² distinct pair types. Direction matters — B-D (backward-on-B ∥ forward-on-D) differs from D-B, since backward carries ~2× the compute. Even a 2-stage pipeline yields eight distinct pair types, each needing its own interleaving.

(b) Variable overlap efficiency — the cycle. Standard partitioners balance the serial execution time of stages, implicitly assuming overlap efficiency is constant. Profiling Qwen3-Next-80B at 256K sequence length on 128 GPUs shows pair C-C gains 41.6% from overlap while D-D gains only 14.0%, because both sides of D-D are compute-dominated with little A2A to hide. Consequently a serial-balanced partition becomes imbalanced after overlap, with a 1.14× higher bottleneck post-overlap cost than an overlap-aware one.

This makes partitioning and overlap scheduling cyclically dependent: the partition determines which pairs form, and the pairs' overlap efficiency determines which partition is best. Partition quality cannot be evaluated without first scheduling and measuring each candidate pair.

<img width="952" height="474" alt="Image" src="https://github.com/user-attachments/assets/d0657a23-7169-4a6c-9a89-699f7eaefd9a" />

(c) Runtime stochasticity. MoE routing sends a fluctuating number of tokens to each expert. Qwen3-Next amplifies this with top-10 / 512 routing: a smaller expected share per expert means higher relative variance, producing ephemeral idle slots no static schedule can predict.

Enabling observation: not all rank-local work is on the critical path. Tessera splits tasks into backbone tasks (define stage latency) and movable tasks (timing-flexible, chiefly weight-gradient Wgrad, whose only hard deadline is the end of the iteration). Movable tasks can be deferred into idle slots.

┌─────────────────────────────────────────────────┬─────────────────────────────────────────────────────────────────┐
│                  Core problem                   │                       Tessera design idea                       │
├─────────────────────────────────────────────────┼─────────────────────────────────────────────────────────────────┤
│ Fixed templates cannot serve asymmetric pair    │ Decompose each pair into task DAGs and synthesize a schedule    │
│ types                                           │ per combination                                                 │
├─────────────────────────────────────────────────┼─────────────────────────────────────────────────────────────────┤
│ Overlap efficiency varies 3×, so serial         │ Balance measured post-overlap cost, not serial arithmetic       │
│ balancing is the wrong objective                │                                                                 │
├─────────────────────────────────────────────────┼─────────────────────────────────────────────────────────────────┤
│ Partition and overlap are cyclically dependent  │ Profile-then-partition: enumerate candidates → schedule →       │
│                                                 │ measure on hardware → then select                               │
├─────────────────────────────────────────────────┼─────────────────────────────────────────────────────────────────┤
│ Analytical overlap models deviate from silicon  │ Ground-truth profiling on a small reference device group        │
├─────────────────────────────────────────────────┼─────────────────────────────────────────────────────────────────┤
│ Routing variation creates unpredictable bubbles │ Dynamic Bubble Optimizer fills predicted slots with movable     │
│                                                 │ tasks                                                           │
└─────────────────────────────────────────────────┴─────────────────────────────────────────────────────────────────┘

---
3 System Design

Tessera fixes the high-level pipeline template and co-optimizes partitioning and overlap scheduling within it. Two layers, separated by a hard boundary: an offline Static Planner for structural heterogeneity, and an online Dynamic Bubble Optimizer (DBO) for runtime stochasticity. DBO never changes partition boundaries or backbone task order.

Static Planning (three steps)

Step 1 — Overlap graph and candidates. The template fixes an overlap graph whose nodes are virtual stages and whose edges are overlap opportunities (each edge recording pass directions). Only which layers each node carries is open. Starting from a serial-balanced baseline, Tessera generates a per-stage candidate set by perturbing chunk boundaries within a bounded neighborhood.

Step 2 — Schedule and profile. Each candidate pair is decomposed into two task DAGs (ATTN, MLP, DISPATCH, COMBINE, WGRAD, each with a duration, a resource type — Comp or Comm — and dependencies). An event-driven list scheduler minimizes the pair's makespan using backbone-first gap-fit placement and conditional deferral of movable tasks. The synthesized schedule is then executed on a 64-GPU reference device group to measure the true post-overlap cost.

Physical profiling is necessary because analytical estimates overshoot by ~5%: EP communication kernels reserve ~20 SMs, causing a stable 10–20% slowdown when overlapped with compute. Since the solver relies on relative comparisons, a few misestimates change the selected partition. Each distinct chunk-pair is profiled only once and cached.

Step 3 — Partition selection. With all costs measured, selection becomes a MILP that minimizes the bottleneck (max) post-overlap edge cost, subject to consistency, contiguous layer ranges, topology, and memory constraints. Edge-based modeling reduces variables from O(N · K^vp) to O(N · vp · K²), and profile-guided pruning shrinks the problem 3–5×.

Dynamic Bubble Optimization

DBO reuses routing metadata already available before dispatch: within each EP group, ranks exchange scalar per-expert token counts on the existing dispatch path, so there is no cluster-wide synchronization. Deviations from the profiled schedule predict where idle slots form. Slot locations are pre-annotated statically, and their sizes are computed asynchronously several steps ahead (token distribution resolves during forward dispatch). A greedy policy then fills each slot from a bounded pool of movable tasks, choosing the tightest duration fit and smallest deadline slack — under 10 µs total. Any task whose slack nears zero is preemptively injected, prioritizing correctness.

---
4 Experiments and Results

Setup. Alibaba production clusters (NVIDIA Hopper, RoCE) with two baselines: an internal optimized Megatron-LM fork already containing interleaved 1F1B, native overlap, and tuned FLOPs-balanced partitioning (up to 12,288 GPUs, isolating incremental benefit); and Megatron-Core MoE v0.16.1 with public recipes on 256 GPUs for reproducibility.

End-to-end. Tessera improves MFU by 20–33% across five production workloads from 4,096 to 12,288 GPUs, reaching 39% MFU on a trillion-parameter model. (Table 2)

Phased deployment. A two-week trace of the 8,192-GPU Qwen3-XL job shows two step changes: the Static Planner gave ~13% (decomposed into ~9% load balancing and ~4% latency hiding — balancing dominates), and enabling DBO raised MFU to 39% while visibly flattening the trace, reducing variability rather than only the mean. (Figure 9)

Gap to ideal. On Qwen3-Next-L, 73% of EP communication is hidden, leaving 17% of iteration time as exposed EP plus pipeline idle — over 40% of it from warmup/cooldown, which is structural to interleaved 1F1B. Qwen3-Next-M hides only 26%, because fewer parameters per expert yield shorter GEMMs relative to A2A. The paper states the boundary plainly: gains scale with the workload's compute-to-communication ratio. (Figure 10)

Versus SOTA. On 256 GPUs, Tessera beats the internal baseline by 1.27× / 1.24× / 1.13× on Qwen3-235B, DeepSeek-V3, and Nemotron-3 Super. Against Megatron-Core it is 1.24× on Qwen3-235B and comparable on the non-Qwen models — Megatron-Core's A2A overlap removes most exposed EP but leaves PP wait intact, whereas Tessera cuts PP wait by ~90% using DBO-deferred Wgrad. (Figure 11)

Co-optimization ablation (the causal proof). With a mock runtime and DBO disabled: without overlap, the serial-balanced Partition A wins (+13.0% vs +12.0%); with overlap enabled, the overlap-aware Partition B wins decisively (+39.1% vs +30.4%). The rank reversal — not just a higher number — confirms that partitioning must be co-optimized with overlap scheduling. (Figure 12)

Component ablations. The scheduling heuristics land within ~1% of an ILP reference while finishing in under a minute versus hours-to-days (Figure 13). DBO cuts iteration time 4.4–5.4%; an always-keep control shows its monitoring machinery costs only ~1%, and peak memory rises 2–4 points from deferred Wgrad (Table 3). Planning is a one-time offline cost: <1 hour of profiling on 64 GPUs, <5 s to solve (Table 4).

---
Summary

Heterogeneous MoE architectures invalidate the assumption that overlap efficiency is constant, turning partitioning and overlap scheduling into a cyclic dependency that stratified optimization cannot resolve. Tessera's answer is to replace estimation with measurement: enumerate candidate partitions, physically profile a synthesized overlap schedule for each candidate pair, then select the partition minimizing the worst measured post-overlap cost — with a lightweight runtime layer absorbing routing stochasticity. In production since April 2025 for Qwen3 and Qwen3-Next, it delivers 20–33% throughput gains at up to 12,288 GPUs while preserving bit-identical gradients.
